File: packages/template-php/src/Ast.php
<?php

declare(strict_types=1);

namespace Polyspec\Template;

/**
 * AST node builders (docs/spec/ast.md). Nodes are nested arrays; spans are byte offsets.
 */
final class Ast
{
    public const LOOP_META_FIELDS = ['index_', 'key_', 'value_', 'last_', 'first_', 'size_'];

    public static function isLoopMetaField(string $name): bool
    {
        return in_array($name, self::LOOP_META_FIELDS, true);
    }

    /** @return array<string, mixed> */
    public static function text(string $value, int $start, int $end): array
    {
        return ['type' => 'Text', 'value' => $value, 'span' => [$start, $end]];
    }

    /**
     * @param array<string, mixed> $expr
     * @return array<string, mixed>
     */
    public static function echo(array $expr, int $start, int $end): array
    {
        return ['type' => 'Echo', 'expr' => $expr, 'span' => [$start, $end]];
    }

    /**
     * @param array<string, mixed> $expr
     * @return array<string, mixed>
     */
    public static function set(string $name, array $expr, int $start, int $end): array
    {
        return ['type' => 'Set', 'name' => $name, 'expr' => $expr, 'span' => [$start, $end]];
    }

    /** @return array<string, mixed> */
    public static function include(string $path, int $start, int $end): array
    {
        return ['type' => 'Include', 'path' => $path, 'span' => [$start, $end]];
    }

    /**
     * @param list<array{name: string, expr: array<string, mixed>}> $scope
     * @return array<string, mixed>
     */
    public static function block(?string $id, ?string $path, array $scope, int $start, int $end): array
    {
        return ['type' => 'Block', 'id' => $id, 'path' => $path, 'scope' => $scope, 'span' => [$start, $end]];
    }

    /** @return array<string, mixed> */
    public static function literal(string $kind, mixed $value, int $start, int $end): array
    {
        return ['type' => 'Literal', 'kind' => $kind, 'value' => $value, 'span' => [$start, $end]];
    }

    /** @return array<string, mixed> */
    public static function variable(string $name, int $start, int $end): array
    {
        return ['type' => 'Var', 'name' => $name, 'span' => [$start, $end]];
    }

    /** @return array<string, mixed> */
    public static function loopMeta(string $loop, string $field, int $start, int $end): array
    {
        return ['type' => 'LoopMeta', 'loop' => $loop, 'field' => $field, 'span' => [$start, $end]];
    }

    /**
     * @param array<string, mixed> $object
     * @return array<string, mixed>
     */
    public static function member(array $object, string $key, int $start, int $end): array
    {
        return ['type' => 'Member', 'object' => $object, 'key' => $key, 'span' => [$start, $end]];
    }

    /**
     * @param array<string, mixed> $object
     * @param array<string, mixed> $index
     * @return array<string, mixed>
     */
    public static function index(array $object, array $index, int $start, int $end): array
    {
        return ['type' => 'Index', 'object' => $object, 'index' => $index, 'span' => [$start, $end]];
    }

    /**
     * @param list<array<string, mixed>> $args
     * @return array<string, mixed>
     */
    public static function call(string $name, array $args, int $start, int $end): array
    {
        return ['type' => 'Call', 'name' => $name, 'args' => $args, 'span' => [$start, $end]];
    }

    /**
     * @param array<string, mixed> $operand
     * @return array<string, mixed>
     */
    public static function unary(string $op, array $operand, int $start, int $end): array
    {
        return ['type' => 'Unary', 'op' => $op, 'operand' => $operand, 'span' => [$start, $end]];
    }

    /**
     * @param array<string, mixed> $left
     * @param array<string, mixed> $right
     * @return array<string, mixed>
     */
    public static function binary(string $op, array $left, array $right, int $start, int $end): array
    {
        return ['type' => 'Binary', 'op' => $op, 'left' => $left, 'right' => $right, 'span' => [$start, $end]];
    }

    /**
     * @param array<string, mixed> $test
     * @param array<string, mixed>|null $then
     * @param array<string, mixed> $else
     * @return array<string, mixed>
     */
    public static function ternary(array $test, ?array $then, array $else, int $start, int $end): array
    {
        return ['type' => 'Ternary', 'test' => $test, 'then' => $then, 'else' => $else, 'span' => [$start, $end]];
    }

    /**
     * @param array<string, mixed> $expr
     * @return array<string, mixed>
     */
    public static function spread(array $expr, int $start, int $end): array
    {
        return ['type' => 'Spread', 'expr' => $expr, 'span' => [$start, $end]];
    }

    /**
     * @param list<array<string, mixed>> $items
     * @return array<string, mixed>
     */
    public static function listLiteral(array $items, int $start, int $end): array
    {
        return ['type' => 'List', 'items' => $items, 'span' => [$start, $end]];
    }

    /**
     * @param list<array<string, mixed>> $entries
     * @return array<string, mixed>
     */
    public static function mapLiteral(array $entries, int $start, int $end): array
    {
        return ['type' => 'Map', 'entries' => $entries, 'span' => [$start, $end]];
    }

    /**
     * Serializes an AST to JSON: number literals as JSON numbers, null fields present, lists as arrays.
     *
     * @param array<string, mixed>|list<mixed> $node
     */
    public static function toJson(mixed $node): string
    {
        if ($node === null) {
            return 'null';
        }
        if (is_bool($node)) {
            return $node ? 'true' : 'false';
        }
        if (is_int($node)) {
            return (string) $node;
        }
        if (is_float($node)) {
            if (Value\Number::isInteger($node) && abs($node) <= Value\Number::MAX_SAFE) {
                return (string) (int) $node;
            }

            return (string) json_encode($node);
        }
        if (is_string($node)) {
            return (string) json_encode($node, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        }
        if (array_is_list($node)) {
            return '[' . implode(',', array_map([self::class, 'toJson'], $node)) . ']';
        }
        $parts = [];
        foreach ($node as $key => $value) {
            $parts[] = json_encode((string) $key) . ':' . self::toJson($value);
        }

        return '{' . implode(',', $parts) . '}';
    }
}

File: packages/template-go/ast/ast.go
// Package ast defines the AST nodes of docs/spec/ast.md and their JSON form.
package ast

// Span is a byte range [start, end) of the template source.
type Span [2]int

// Template is the root node.
type Template struct {
	Type string `json:"type"`
	Name string `json:"name"`
	Body []Node `json:"body"`
}

// Node is a statement node.
type Node interface{ node() }

// Expr is an expression node.
type Expr interface{ expr() }

// Text node.
type Text struct {
	Type  string `json:"type"`
	Value string `json:"value"`
	Span  Span   `json:"span"`
}

// Echo node.
type Echo struct {
	Type string `json:"type"`
	Expr Expr   `json:"expr"`
	Span Span   `json:"span"`
}

// IfBranch is one branch of an If node.
type IfBranch struct {
	Test Expr   `json:"test"`
	Body []Node `json:"body"`
	Span Span   `json:"span"`
}

// If node.
type If struct {
	Type     string      `json:"type"`
	Branches []*IfBranch `json:"branches"`
	Else     []Node      `json:"else"`
	Span     Span        `json:"span"`
}

// For node.
type For struct {
	Type  string `json:"type"`
	Name  string `json:"name"`
	Iter  Expr   `json:"iter"`
	Body  []Node `json:"body"`
	Empty []Node `json:"empty"`
	Span  Span   `json:"span"`
}

// Set node.
type Set struct {
	Type string `json:"type"`
	Name string `json:"name"`
	Expr Expr   `json:"expr"`
	Span Span   `json:"span"`
}

// Include node.
type Include struct {
	Type string `json:"type"`
	Path string `json:"path"`
	Span Span   `json:"span"`
}

// ScopeItem is one scope argument of a block tag.
type ScopeItem struct {
	Name string `json:"name"`
	Expr Expr   `json:"expr"`
}

// Block node.
type Block struct {
	Type  string       `json:"type"`
	ID    *string      `json:"id"`
	Path  *string      `json:"path"`
	Scope []*ScopeItem `json:"scope"`
	Span  Span         `json:"span"`
}

// IfBlock node.
type IfBlock struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	Body []Node `json:"body"`
	Else []Node `json:"else"`
	Span Span   `json:"span"`
}

func (*Text) node()    {}
func (*Echo) node()    {}
func (*If) node()      {}
func (*For) node()     {}
func (*Set) node()     {}
func (*Include) node() {}
func (*Block) node()   {}
func (*IfBlock) node() {}

// Literal node; Value is nil, bool, float64 or string.
type Literal struct {
	Type  string `json:"type"`
	Kind  string `json:"kind"`
	Value any    `json:"value"`
	Span  Span   `json:"span"`
}

// Var node.
type Var struct {
	Type string `json:"type"`
	Name string `json:"name"`
	Span Span   `json:"span"`
}

// LoopMeta node.
type LoopMeta struct {
	Type  string `json:"type"`
	Loop  string `json:"loop"`
	Field string `json:"field"`
	Span  Span   `json:"span"`
}

// Member node.
type Member struct {
	Type   string `json:"type"`
	Object Expr   `json:"object"`
	Key    string `json:"key"`
	Span   Span   `json:"span"`
}

// MemberCall invokes a method on an assigned object.
type MemberCall struct {
	Type   string `json:"type"`
	Object Expr   `json:"object"`
	Method string `json:"method"`
	Args   []Expr `json:"args"`
	Span   Span   `json:"span"`
}

// ClassCall invokes a declared logical class function.
type ClassCall struct {
	Type      string `json:"type"`
	ClassName string `json:"className"`
	Method    string `json:"method"`
	Args      []Expr `json:"args"`
	Span      Span   `json:"span"`
}

// Index node.
type Index struct {
	Type   string `json:"type"`
	Object Expr   `json:"object"`
	Index  Expr   `json:"index"`
	Span   Span   `json:"span"`
}

// Call node.
type Call struct {
	Type string `json:"type"`
	Name string `json:"name"`
	Args []Expr `json:"args"`
	Span Span   `json:"span"`
}

// Unary node.
type Unary struct {
	Type    string `json:"type"`
	Op      string `json:"op"`
	Operand Expr   `json:"operand"`
	Span    Span   `json:"span"`
}

// Binary node.
type Binary struct {
	Type  string `json:"type"`
	Op    string `json:"op"`
	Left  Expr   `json:"left"`
	Right Expr   `json:"right"`
	Span  Span   `json:"span"`
}

// Ternary node; Then is nil for the elvis form.
type Ternary struct {
	Type string `json:"type"`
	Test Expr   `json:"test"`
	Then Expr   `json:"then"`
	Else Expr   `json:"else"`
	Span Span   `json:"span"`
}

// Spread entry of a list or map literal.
type Spread struct {
	Type string `json:"type"`
	Expr Expr   `json:"expr"`
	Span Span   `json:"span"`
}

// List literal; Items holds Expr or *Spread.
type List struct {
	Type  string `json:"type"`
	Items []any  `json:"items"`
	Span  Span   `json:"span"`
}

// MapEntry of a map literal.
type MapEntry struct {
	Key   Expr `json:"key"`
	Value Expr `json:"value"`
}

// Map literal; Entries holds *MapEntry or *Spread.
type Map struct {
	Type    string `json:"type"`
	Entries []any  `json:"entries"`
	Span    Span   `json:"span"`
}

func (*Literal) expr()  {}
func (*Var) expr()      {}
func (*LoopMeta) expr() {}
func (*Member) expr()   {}
func (*MemberCall) expr() {}
func (*ClassCall) expr() {}
func (*Index) expr()    {}
func (*Call) expr()     {}
func (*Unary) expr()    {}
func (*Binary) expr()   {}
func (*Ternary) expr()  {}
func (*List) expr()     {}
func (*Map) expr()      {}

// LoopMetaFields are the fields of EXP-21.
var LoopMetaFields = map[string]bool{"index_": true, "key_": true, "value_": true, "last_": true, "first_": true, "size_": true}

// SpanOf returns the span of an expression node.
func SpanOf(e Expr) Span {
	switch n := e.(type) {
	case *Literal:
		return n.Span
	case *Var:
		return n.Span
	case *LoopMeta:
		return n.Span
	case *Member:
		return n.Span
	case *MemberCall:
		return n.Span
	case *ClassCall:
		return n.Span
	case *Index:
		return n.Span
	case *Call:
		return n.Span
	case *Unary:
		return n.Span
	case *Binary:
		return n.Span
	case *Ternary:
		return n.Span
	case *List:
		return n.Span
	case *Map:
		return n.Span
	}
	return Span{}
}
